iib/workers/tasks/build_merge_index_image.py

```diff
@@ -0,0 +1,366 @@
+        '--generate',
+        '--binary-image',
+        binary_image,
+        '--from-index',
+        from_index,
+        '--bundles',
+        ','.join(bundles),
+    ]
+    # TODO catch error of default channel deletion
+    with set_registry_token(overwrite_from_index_token, from_index):
+        run_cmd(
+            cmd, {'cwd': base_dir}, exc_msg='Failed to deprecate the bundles',
+        )
+
+    _add_ocp_label_to_index(ocp_version, base_dir, 'index.Dockerfile')
+
+
+@app.task
+@request_logger
+def handle_merge_request(
+    binary_image,
+    source_from_index,
+    target_index,
+    deprecation_list,
+    request_id,
+    overwrite_from_index=False,
+    overwrite_from_index_token=None,
+):
+    """
+    Coordinate the work needed to merge old (N) index image with new (N+1) index image.
+
+    :param str binary_image: the pull specification of the container image where the opm binary
+        gets copied from.
+    :param str source_from_index: pull specification to be used as base for building the new
+        index image.
+    :param str target_index: pull specification of content stage index image for the
+        corresponding target index image.
+    :param list deprecation_list: List of deprecated bundles for the target index image tag.
+    :param int request_id: the ID of the IIB build request.
+    :param bool overwrite_from_index: if True, overwrite the input ``target_index`` with
+        the built index image.
+    :param str overwrite_from_index_token: the token used for overwriting the input
+        ``target_index`` image. This is required for non-privileged users to use
+        ``overwrite_from_index``. The format of the token must be in the format "user:password".
+    :raises IIBError: if the index image merge fails.
+    """
+    _cleanup()
+    prebuild_info = _prepare_request_for_build(
+        binary_image,
+        request_id,
+        overwrite_from_index_token=overwrite_from_index_token,
+        source_from_index=source_from_index,
+        target_index=target_index,
+    )
+    _update_index_image_build_state(request_id, prebuild_info)
+
+    with tempfile.TemporaryDirectory(prefix='iib-') as temp_dir:
+        set_request_state(request_id, 'in_progress', 'Getting bundles present in the index images')
+        log.info('Getting bundles present in the source index image')
+        source_index_bundles = _get_present_bundles(source_from_index, temp_dir)
+        log.info('Getting bundles present in the target index image')
+        target_index_bundles = _get_present_bundles(target_index, temp_dir)
+
+        arches = list(prebuild_info['arches'])
+        arch = 'amd64' if 'amd64' in arches else arches[0]
+        _create_intermediate_index_image(
+            binary_image,
+            source_from_index,
+            prebuild_info['target_ocp_version'],
+            temp_dir,
+            request_id,
+            arch,
+        )
+
+        missing_bundles = _add_bundles_missing_in_source(
```

Comment:
Instead of calling opm_index_add twice in this function and in _create_intermediate_index_image, what's stopping us from finding the missing bundles first and then simply calling opm_index_add once along with the missing bundles to form the intermediate index image. Then simply call build_image once for each arch on the index image. Wouldn't this approach reduce complexity?

Reply:
Yeah, I noticed that the process could be optimized, but decided to implement it as it was described in the design doc in case I was missing some information.

I have simplified the code according to your suggestion.